src/masternodes/masternodes.cpp

```diff
@@ -917,73 +917,129 @@ CAmount CCollateralLoans::precisionRatio() const
-        auto amount = MultiplyAmounts(price, col.second);
-        if (price > COIN && amount < col.second)
-            return Res::Err("Value/price too high (%s/%s)", GetDecimaleString(col.second), GetDecimaleString(price));
-        amount = MultiplyAmounts(collToken->factor, amount);
-        auto prevCollaterals = ret.totalCollaterals;
-        ret.totalCollaterals += amount;
-        if (prevCollaterals > ret.totalCollaterals)
-            return Res::Err("Exceed maximum collateral");
-        ret.collaterals.push_back({col.first, amount});
+    auto priceRecordIndex = useNextPrice ? 1 : 0;
+    auto price = priceFeed.val->priceRecord[priceRecordIndex];
+    if (price <= 0)
+        return Res::Err("Negative price (%s/%s)", tokenSymbol, currency);
+
+    return ResVal<CAmount>(price, Res::Ok());
+}
+
+Res PopulateLoansData(CCustomCSView *view, CCollateralLoans& result, CVaultId const& vaultId, 
+    uint32_t height, int64_t blockTime, bool useNextPrice, bool requireLivePrice) {
+    if (!view)
+        return Res::Err("Unexpected call without chain state");
+    
+    auto loanTokens = view->GetLoanTokens(vaultId);
+    if (!loanTokens)
+        return Res::Ok();
+
+    for (const auto& loan : loanTokens->balances) {
+        auto loanTokenId = loan.first;
+        auto loanTokenAmount = loan.second;
+
+        auto token = view->GetLoanSetLoanTokenByID(loanTokenId);
+        if (!token)
+            return Res::Err("Loan token with id (%s) does not exist!", loanTokenId.ToString());
+
+        auto rate = view->GetInterestRate(vaultId, loanTokenId);
+        if (!rate)
+            return Res::Err("Cannot get interest rate for token (%s)!", token->symbol);
+        if (rate->height > height)
+            return Res::Err("Trying to read loans in the past");
+
+        auto priceResult = GetValidatedIntervalPrice(view, token->fixedIntervalPriceId, useNextPrice, requireLivePrice);
+        if (!priceResult.ok)
+            return std::move(priceResult);
+        auto price = priceResult.val.get();
+
+        LogPrint(BCLog::LOAN,"\t\t%s()->for_loans->%s->", __func__, token->symbol); /* Continued */
+        
+        auto value = loanTokenAmount + TotalInterest(*rate, height);
+        auto amountInCurrency = MultiplyAmounts(price, value);
+        
+        if (price > COIN && amountInCurrency < value)
+            return Res::Err("Value/price too high (%s/%s)", GetDecimaleString(value), GetDecimaleString(price));
+
+        auto prevLoans = result.totalLoans;
+        result.totalLoans += amountInCurrency;
+        
+        if (prevLoans > result.totalLoans)
+            return Res::Err("Exceeded maximum loans");
+
+        result.loans.push_back({loanTokenId, amountInCurrency});
     }
+    return Res::Ok();
+}
+
+Res PopulateCollateralData(CCustomCSView *view, CCollateralLoans& result, CVaultId const& vaultId,
```

Comment:
Static, move over `GetLoanCollaterals`, use `CCustomCSView& view`

Reply:
Same as the other.